### ISSUE TITLE
aws-ecs: wrong cfnspec for EFS volume configuration (casing issue)

### ISSUE THREAD

**Issue by lavie** [aws/aws-cdk#15025]
CDK currently generates two resources with casing mistakes when `EFSVolumeConfiguration` is specified for task definitions.

`EfsVolumeConfiguration` should be 
`EFSVolumeConfiguration`

`FileSystemId` should be
`FilesystemId`

The corresponding test is also wrong: 
https://github.com/aws/aws-cdk/blob/23986d70c5cd69ce212b5ffdc1bcf059f438f15b/packages/%40aws-cdk/aws-ecs/test/ec2/ec2-task-definition.test.ts#L1192

### Reproduction Steps

The following (incomplete) snippet recreates the issue:

```
      const efs = new FileSystem(this, 'etcdata', {
        vpc: service.cluster.vpc,
      });

      taskDefinition.addVolume({
        name: 'etcdata',
        efsVolumeConfiguration: {
          fileSystemId: efs.fileSystemId
        },
      });
```

This is based on the `ApplicationLoadBalancedFargateService` pattern.

The above generates this CFN JSON:

```
        "Volumes": [
          {
            "Name": "etcdata",
            "EfsVolumeConfiguration": {
              "FileSystemId": {
                "Ref": "etcdata80702D7D"
              }
            }
          }
        ]
```

Which then yields CFN errors:

```message: #/Volumes/0: extraneous key [EfsVolumeConfiguration] is not permitted```

And if I fix the above to use the correct case manually, then later it uncovers the 2nd problem:

```
 #/Volumes/0/EFSVolumeConfiguration: required key [FilesystemId] not found
```

Fixing both manually causes the template to be deployed successfully.

### Environment

  - **CDK CLI Version  : 1.107.0 (build 52c4434)**
  - **Framework Version: ??**
  - **Node.js Version: v16.2.0** 
  - **OS               : OSX**
  - **Language (Version): typescript 4.3.2 -->

--- 

This is :bug: Bug Report

**Comment by mteichtahl** [comment]
I have also experienced this bug with the same symptoms and reproducibility. This is a significant issue for those needing to mount EFS volumes to a garbage container.

**Comment by MrArnoldPalmer** [comment]
TLDR: This validation message is a warning, if you're stack is failing to deploy it's because of another problem. If it's unclear in the cloudformation console check out ECS console and look at your stopped tasks during the stack deploy to find the error messaging. We will fix this in 2.0!

So we know about the incorrect casing on these properties, but it doesn't cause the stack deploy to fail. When reproducing this I got this message in the CFN console as well, but it didn't fail my deploy. What is most likely happening (what happened to me and took me a bit to figure out) is that you're missing the `fs.connections.allowDefaultPortFrom(service);` which causes your tasks to fail the healthcheck with `ResourceInitializationError: failed to invoke EFS utils`. It will keep redeploying these containers for like 45 minutes before timing out and the stack deploy fails. You can look in the ECS console during the deploy and find the service, then click the `tasks` tab, then click the `stopped` filter. If you have stopped tasks there, you can click on them to see the details and this is where you'll see the error.

<img width="610" alt="Screen Shot 2021-06-09 at 7 22 18 PM" src="https://user-images.githubusercontent.com/7221111/121454833-5d075a00-c958-11eb-8940-4d6cc1e6d929.png">

Here is a minimal stack that I was able to deploy.
<details>
  <summary>Click Me!</summary>
  
```typescript
export class Issue15025Stack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, "Vpc");
    const cluster = new ecs.Cluster(this, "Cluster", {
      vpc,
    });
    const fs = new efs.FileSystem(this, "FileSystem", {
      vpc,
    });

    const taskDefinition = new ecs.TaskDefinition(this, "TaskDefinition", {
      cpu: "256",
      memoryMiB: "512",
      compatibility: ecs.Compatibility.FARGATE,
    });

    taskDefinition.addContainer("Container", {
      image: ecs.ContainerImage.fromAsset(
        path.resolve(__dirname, "..", "container")
      ),
    });

    taskDefinition.addVolume({
      name: "TaskFs",
      efsVolumeConfiguration: {
        fileSystemId: fs.fileSystemId,
      },
    });

    const service = new ecs.FargateService(this, "Service", {
      cluster,
      taskDefinition,
    });

    fs.connections.allowDefaultPortFrom(service);
    // The code that defines your stack goes here
  }
}
```
</details>

Long story, but what happened is a contributor edited the CloudFormation spec manually to add these properties so they could use EFS with ECS. We ingest the new cloudformation spec once a week, but sometimes there are bugs that cause this to take a little longer and without knowing that this was automated, they made this change to unblock themselves and we mistakenly accepted the PR. This is where the properties with the bad casing were ingested. https://github.com/aws/aws-cdk/pull/8467

Tests showed that stacks deployed correctly with this initial PR.

Then, a couple weeks later when we ingested the new spec with these properties included, we saw that they were in fact incorrectly cased. So if we changed the casing, this would break users who already had stacks deployed using this casing. "Break" as in break backwards compatibility because they would have to update their stack code to use the correct property names. Instead [we patched the CFN spec](https://github.com/aws/aws-cdk/pull/10483) so we wouldn't break users.

Regardless we should fix this, and we have a v2.0 in progress that gives us the opportunity to do so. I'm keeping this open for @SoManyHs and @madeline-k to track removing this patch for 2.0.

**Comment by mteichtahl** [comment]
A hack approach to remove those warnings

```typescript
        /** max hackiness to handle https://github.com/aws/aws-cdk/issues/15025 */
        const volumes = [optSonarDataVolume, optSonarTempVolume, optSonarExtVolume, optSonarLogVolume]
        const cfnTD = (this.fargate.taskDefinition.node.defaultChild as CfnTaskDefinition);

        for (let i = 0; i < volumes.length; i++) {
            this.fargate.taskDefinition.addVolume(volumes[i])
            const { fileSystemId, transitEncryption, rootDirectory, ...rest } = volumes[i].efsVolumeConfiguration!
            cfnTD.addPropertyOverride(`Volumes.${i}.EFSVolumeConfiguration`, {
                ...rest,
                FilesystemId: fileSystemId,
                TransitEncryption: transitEncryption,
                RootDirectory: rootDirectory
            });
            cfnTD.addPropertyDeletionOverride(`Volumes.${i}.EfsVolumeConfiguration`);
        }
        /** end max hackiness */
```

**Comment by namedgraph** [comment]
@MrArnoldPalmer what do you mean with "if you're stack is failing to deploy it's because of another problem"? My stack fails to deploy after I started using `efsVolumeConfiguration()` on my task definition:
```
Resource template validation failed for resource LDHTaskDefinitionDF5ADF89 as the template has invalid properties.
Please refer to the resource documentation to fix the template.
Properties validation failed for resource LDHTaskDefinitionDF5ADF89 with message: #/Volumes/2: extraneous key [EfsVolumeConfiguration] is not permitted #/Volumes/5: extraneous key [EfsVolumeConfiguration] is not permitted
```

**Comment by mteichtahl** [comment]
It may appear to be the cause though these are warnings. They won’t cause your stack to fail. 

However, if you depend on the volume it’s likely to not be deployed causing “other” issues. 

Please note: My working hours may not be your working hours. Please do not feel obligated to reply outside of your normal work schedule!

Sent on the go - please excuse mistakes 

On Jun 18, 2021, at 3:06 AM, Martynas Jusevičius ***@***.***> wrote:

﻿
@MrArnoldPalmer what do you mean with "if you're stack is failing to deploy it's because of another problem"? My stack fails to deploy after I started using efsVolumeConfiguration() on my task definition:

Resource template validation failed for resource LDHTaskDefinitionDF5ADF89 as the template has invalid properties.
Please refer to the resource documentation to fix the template.
Properties validation failed for resource LDHTaskDefinitionDF5ADF89 with message: #/Volumes/2: extraneous key [EfsVolumeConfiguration] is not permitted #/Volumes/5: extraneous key [EfsVolumeConfiguration] is not permitted

**Comment by namedgraph** [comment]
@mteichtahl well why use the volume if not to depend on it? :) We want to use an EFS volume to persist DB data. If that doesn't work, of course our DB container will fail.

Let me ask this way: is EFS volume support for task definitions currently broken in CDK?

**Comment by namedgraph** [comment]
I checked my ECS service logs and what I see is:
```
(service LinkedDataHubStack-EC2Service904511EF-5bXvSMuNKbl9) was unable to place a task because no container instance met all of its requirements. Reason: No Container Instances were found in your cluster.
```

**Comment by MrArnoldPalmer** [comment]
@namedgraph what I mean is that I was able to deploy a stack successfully using efsVolumeConfiguration, so my conclusion was that this message is a warning but ultimately it doesn't affect stack deployment.

The error in your service logs looks related to incorrect cluster/instance configuration. `fs.connections.allowDefaultPortFrom(service);` is used to give the service access to the filesystem, but I think you would also need to give your ec2 instances access as well if you haven't. Just a suggestion but it could be something else. Check out [these docs](https://aws.amazon.com/premiumsupport/knowledge-center/ecs-container-instance-requirement-error/) and see if any of that information is relevant.

> Let me ask this way: is EFS volume support for task definitions currently broken in CDK?

I don't think so. At least if it is I don't believe that this is the cause of it.

> However, if you depend on the volume it’s likely to not be deployed causing “other” issues.

@mteichtahl I'm curious exactly what you mean by this as well. I didn't see any problem with the volume on my deploy but I'm not super experienced with EFS still.

**Comment by namedgraph** [comment]
@MrArnoldPalmer so OK, maybe there is another error but it's got to be EFS-related, because if I replace `efsVolumeConfiguration()` with `host()` my stack deploys just fine.

But I haven't been able to find that specific error yet.

**Comment by MrArnoldPalmer** [comment]
@namedgraph makes sense to me. Are you able to share your code or a minimal reproduction that has the same error?

**Comment by namedgraph** [comment]
@MrArnoldPalmer if I had a good example, it would probably take me less time to fix my problem than to create a reproduction :)
I can see I'm missing some permissions:
https://gist.github.com/cajames/3daec680b1101c8358e2ff30dfadd52a?fbclid=IwAR0UGQ3TfwIyWvsi6R64zH4TUNFgc-1iEdNNAajRpCsCffx0nWMCbpIRrs4#file-fargate-efs-cdk-stack-ts-L176

**Comment by namedgraph** [comment]
I added `ec2Service.getService().getConnections().allowFrom()/allowTo()` but the deployment still fails.

Though I was finally able to find the actual error in the DB task logs:
```
Error response from daemon: create ecs-LinkedDataHubStackLDHTaskDefinitionF106B511-162-FusekiAdminDataVolume-e69dae89abd09e9de901: VolumeDriver.Create: mounting volume failed: b'mount.nfs4: mounting fs-468514f2.efs.us-east-1.amazonaws.com:/var/fuseki/data/admin failed, reason given by server: No such file or directory'
```

**Comment by mteichtahl** [comment]
Can you share the synthesised cloud formation for the efs mount point ?

Please note: My working hours may not be your working hours. Please do not feel obligated to reply outside of your normal work schedule!

Sent on the go - please excuse mistakes 

On Jun 19, 2021, at 3:30 AM, Martynas Jusevičius ***@***.***> wrote:

﻿
I added ec2Service.getService().getConnections().allowFrom()/allowTo() but the deployment still fails.

Though I was finally able to find the actual error in the DB task logs:

Error response from daemon: create ecs-LinkedDataHubStackLDHTaskDefinitionF106B511-162-FusekiAdminDataVolume-e69dae89abd09e9de901: VolumeDriver.Create: mounting volume failed: b'mount.nfs4: mounting fs-468514f2.efs.us-east-1.amazonaws.com:/var/fuseki/data/admin failed, reason given by server: No such file or directory'

**Comment by namedgraph** [comment]
These are the relevant parts from the template, I think:
```
  LDHFileSystem0D5F0D0E:
    Type: AWS::EFS::FileSystem
    Properties:
      Encrypted: true
      FileSystemTags:
        - Key: Name
          Value: LinkedDataHubStack/LDHFileSystem
    UpdateReplacePolicy: Retain
    DeletionPolicy: Retain
    Metadata:
      aws:cdk:path: LinkedDataHubStack/LDHFileSystem/Resource

  LDHTaskDefinitionDF5ADF89:
    Type: AWS::ECS::TaskDefinition
    Properties:
      Volumes:
        - Name: FusekiAdminDataVolume
          EfsVolumeConfiguration:
            RootDirectory: /var/fuseki/data/admin
            FileSystemId:
              Ref: LDHFileSystem0D5F0D0E
          MountPoints:
            - ContainerPath: /var/fuseki/data
              ReadOnly: false
              SourceVolume: FusekiAdminDataVolume
```
Am I supposed to create the directories on EFS?

Using CDK 1.102.0.

**Comment by lavie** [comment]
@namedgraph I don't see how this is contributing to the issue I've reported. The team has acknowledged the error and presumably it will be fixed soon, since it's quite a simple fix. If you're having problems deploying even after using the workaround which fixes the original issue then I suggest you open a separate ticket.

**Comment by namedgraph** [comment]
@lavie sure. What exactly is the workaround - this? https://github.com/aws/aws-cdk/issues/15025#issuecomment-858245732

**Comment by lavie** [comment]
yea, it works.

**Comment by epapathom** [comment]
One month later this is still not fixed. Can we expect a solution?

**Comment by nija-at** [comment]
The fix involves a change like this - https://github.com/aws/aws-cdk/commit/dd308b6610a670f0dd91df2476253cffb4cb04a3 - but behind a feature flag to avoid resource replacements on stacks where this already works.

**Comment by desoss** [comment]
Still not solved in v1.116.0 😢

**Comment by jeznag** [comment]
Chiming in that for me the error message did not cause any problems and EFS still mounted correctly to my fargate container.

Here's a gist
https://gist.github.com/jeznag/623132d2a98eff5392acd702702f1ef7

**Comment by v1pz3n** [comment]
The problem persists in 1.122 and 2.0.
Writing in Python is not even overridable.

By the way rewrite in CfnService is better.

**Comment by ScottHoward4** [comment]
> A hack approach to remove those warnings
> 
> ```ts
>         /** max hackiness to handle https://github.com/aws/aws-cdk/issues/15025 */
>         const volumes = [optSonarDataVolume, optSonarTempVolume, optSonarExtVolume, optSonarLogVolume]
>         const cfnTD = (this.fargate.taskDefinition.node.defaultChild as CfnTaskDefinition);
> 
>         for (let i = 0; i < volumes.length; i++) {
>             this.fargate.taskDefinition.addVolume(volumes[i])
>             const { fileSystemId, transitEncryption, rootDirectory, ...rest } = volumes[i].efsVolumeConfiguration!
>             cfnTD.addPropertyOverride(`Volumes.${i}.EFSVolumeConfiguration`, {
>                 ...rest,
>                 FilesystemId: fileSystemId,
>                 TransitEncryption: transitEncryption,
>                 RootDirectory: rootDirectory
>             });
>             cfnTD.addPropertyDeletionOverride(`Volumes.${i}.EfsVolumeConfiguration`);
>         }
>         /** end max hackiness */
> ```

This got rid of the error but still no volumes were created. Anyone else have similar problems?

**Comment by dave-juicelabs** [comment]
In case any one else is using python and having issues with this

```
from aws_cdk import (
    Stack,
    aws_ec2 as ec2,
    aws_ecs as ecs,
    aws_efs as efs
)
from constructs import Construct

class JenkinsHomeStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, vpc: ec2.Vpc, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        filesystem = efs.FileSystem(self, 'jenkins-home', vpc=vpc)
        self.config = ecs.EfsVolumeConfiguration(file_system_id=filesystem.file_system_id)
```

```
from aws_cdk import (
    Duration,
    Stack,
    aws_ec2 as ec2,
    aws_ecs as ecs,
    aws_ecs_patterns as ecs_patterns,
    aws_elasticloadbalancingv2 as elb
)
from constructs import Construct

class JenkinsStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, vpc: ec2.Vpc, jenkinsHomeEfsConfig: ecs.EfsVolumeConfiguration, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        jenkinsCluster = ecs.Cluster(self, 'cluster', vpc=vpc)

        jenkinsService = ecs_patterns.ApplicationLoadBalancedFargateService(self, 'JenkinsMasterService', assign_public_ip=True, 
            cluster=jenkinsCluster, cpu=512, memory_limit_mib=1024, desired_count=1, listener_port=80,
            protocol=elb.ApplicationProtocol.HTTP, task_image_options=ecs_patterns.ApplicationLoadBalancedTaskImageOptions(
                image=ecs.ContainerImage.from_registry('jenkins/jenkins:lts-jdk11'), container_port=8080))
        jenkinsService.target_group.configure_health_check(path='/login', port='8080', timeout=Duration.seconds(30),
            interval=Duration.seconds(60), unhealthy_threshold_count=5, healthy_threshold_count=5)
        jenkinsService.task_definition.add_volume(efs_volume_configuration=jenkinsHomeEfsConfig, name='jenkins-home')
        jenkinsService.task_definition.default_container.add_mount_points(
            ecs.MountPoint(container_path='/var/jenkins_home', source_volume='jenkins-home', read_only=False))

        # HACK: To fix incorrect CloudFormation names generated by aws_cdk as of v2.3.0
        task = jenkinsService.task_definition.node.default_child
        task.add_property_override('Volumes.0.EFSVolumeConfiguration', {'FilesystemId': jenkinsHomeEfsConfig.file_system_id})
        task.add_property_deletion_override('Volumes.0.EfsVolumeConfiguration')
```

**Comment by mathieubrd** [comment]
Any update on this?
Bug still there with CDK v2.8

**Comment by Blufe** [comment]
Me too...

**Comment by hf-qin** [comment]
I can confirm that the bug is still there with CDV v2.12

**Comment by bearrito** [comment]
Still present in 2.24.1

**Comment by FabrizioCafolla** [comment]
> Still present in 2.24.1

I also have the same error but only if I add the parameter: `root_category`

Do you have a workaround?
